[JENKINS-51480] Fixes for environment variables

proski opened the pull request:
Edited: no longer work in progress, should be good to commit.

First commit:
* There are no more warnings about environment variables from the freestyle projects (JENKINS-51480).

Second commit:
* Child jobs in mutliconfiguration projects get the environment variables without any special settings.

Not addressed by this commit, will be part of the JENKINS-56012 work:
* Mutliconfiguration projects still report a lot of warnings about parameters if configurations are tested subsequently. That may actually be a Jenkins bug. Something gets reused, and the warnings are actually incorrect, as the project does define those variables. The amount of the warnings (15 per variable per build) suggests that they come non-build runs (e.g. git invocations).

proski committed:
Don't put environment variables to the build parameters

Take the values for the environment variables directly from StashCause.

This fixes warnings about parameters for freestyle projects. Also, the
environment variables are listed at one place now.

proski committed:
Provide environment variables to downstream builds

The cause would only be StashCause for the the top-level build of a
multi-configuration project.

First make sure it's an AbstractBuild. Non-build runs (e.g. git
invocations) should not need our environment variables (the variables are
substituted explicitly before the command is run).

Then find the root build and check it's cause.

Don't return prematurely to ensure that super.buildEnvironmentFor() is
run unconditionally.

proski commented:
The approach with AbstractBuild#getRootBuild is much cleaner, so I'm using it now. We are not supposed to provide the environment variables to runs that are not builds (e.g. git invocations).

There are still warnings, but only from the second configuration. That is, the top-level build runs, the "Release" configuration runs and completes successfully without warnings, then the "Debug" configuration starts and emits 15 sets of warnings.

The log for the "Debug" build shows 10 git invocations. I believe the warnings are from them and from 5 more non-build runs that are not logged. I guess they reuse something and end us with parameters that they should not have.

jakub-bochenski commented on src/main/java/stashpullrequestbuilder/stashpullrequestbuilder/StashAditionalParameterEnvironmentContributor.java:
```diff
@@ -10,40 +10,32 @@
+            AbstractBuild rootBuild = build.getRootBuild();
+
+            StashCause cause = (StashCause) rootBuild.getCause(StashCause.class);
+            if (cause != null) {
```
minor: can we leave the `cause == null` check at the top, just to make the method simpler?

jakub-bochenski replied:
nvm: I just noticed this is cause from rootBuild and not from the run

proski replied:
I don't want to call `super.buildEnvironmentFor()` conditionally. It's actually an empty function in `EnvironmentContributor`, but I want to be consistent. Either we call it and don't return prematurely, or we leave it out completely.

jakub-bochenski commented:
> Why didn't the warnings stop?

No idea on that, maybe this is something to ask on the mailing list

>    If it always a good idea to check the upstream jobs? Can they have other meanings apart from the child jobs in the matrix?
    Do we want to check the upstream job of the upstream job? How far should we go?
    Maybe AbstractBuild#getRootBuild is the way to go?

I agree, using rootBuild seems more robust.

> Why don't I see any code similar to the second patch in the GitHub and Bitbucket PR builder plugins? I wonder how they work with the multi-configuration projects.

Do you know if they support multi-configuration projects at all?
I don't think https://github.com/jenkinsci/gitlab-plugin supports posting comments at all

jakub-bochenski commented on src/main/java/stashpullrequestbuilder/stashpullrequestbuilder/StashAditionalParameterEnvironmentContributor.java:
```diff
@@ -10,40 +10,32 @@
-        StringParameterValue value = (StringParameterValue) pa.getParameter(key);
-        envs.put(key, getString(value.value, ""));
+    private static void putEnvVar(EnvVars envs, String key, String value) {
+        envs.put(key, getString(value, ""));
```
minior: can you use `Objects.toString(value,"")` instead?

proski replied:
Good point, I'm making that change.

jakub-bochenski commented:
Thanks for all the work done here.

It would be nice to get some automated tests for this with Jenkins Unit Test Harness, though I won't insist, since the existing coverage is rather poor anyway.

jimklimov reviewed (approved):
If nothing else, this looks a lot more compact to read and debug :)

proski commented:
Addressed the comment. Expanded commit texts, described the user visible effects.

I actually feel much more confident about both changes. The first commit resolves JENKINS-51480. The second commit resolves part of JENKINS-56012 (missing variables in child builds).
